### DISCUSSION TITLE
Prefect 3 - Error running task with a custom Docker container

### DISCUSSION THREAD

**Discussion by Lukem76** [PrefectHQ/prefect#16160] [Q&A]
I have a project with a simple python file that is containerized with a Dockerfile.
The dockerfile is:

```Dockerfile
FROM python:3.9-slim

WORKDIR /app

COPY . .

RUN pip install --no-cache-dir -r requirements.txt

CMD ["python", "prefect-test.py"]
```

`prefect-test.py` is a simple file that print a message. If I build the image locally with Docker Desktop and run it, it works. I already pushed it on docker hub in a public repo.

Now I want to start a Prefect server versione 3 (not 2, it's important). I created this Dokerfile:

```Dockerfile
FROM prefecthq/prefect:3-latest

WORKDIR /app

CMD ["prefect", "server", "start", "--host", "0.0.0.0"]
```

I build the image and ran it with:

```bash
docker build -t prefect-server .

docker run -d -p 4200:4200 --name prefect-server prefect-server
```

Then I access the GIU at http://localhost:4200 

Everything works.

Now I want to create a work pool, start an agent and do a deploy. The target is to create a deploy file that run the Dockerimage that run the prefect-test.py file.

I create the work pool with:

```bash
prefect work-pool create --type docker my-work-pool
```

I started a worker:

```bash
prefect worker start --pool docker my-work-pool
```

Now I have to define the deployment. I created a file named deploy_buy.py in a different folder than the root where I have the Dockerfile and prefect-test.py file:

```
from prefect import flow

@flow(log_prints=True)
def buy():
print("Selling securities")

if __name__ == "__main__":
  buy.deploy(
	  name="my-deployment",
	  work_pool_name="my-work-pool",
	  image="docker.io/xxx/image"
  )
```

I copied it from official doc here: [https://docs.prefect.io/v3/deploy/infrastructure-examples/docker][1]

  [1]: https://docs.prefect.io/v3/deploy/infrastructure-examples/docker

If I run the task, I see the message "Buying securities" in logs and at the end I see:
Finished in state Completed()
Process for flow run 'xxx' exited cleanly.
But the file prefect-test.py wasn't invoked, because I don't see any other log.

I used a different folder for deploy file because, if I don't do that, when I run deploy, it gives me an error saying that a Dockerfile is already in that folder. Executing in a different folder, it creates automatically a Dockerfile with this:
```Dockerfile
FROM prefecthq/prefect:3.1.4-python3.10
COPY . /opt/prefect/deploy/
WORKDIR /opt/prefect/deploy/
```

This is not clear, because that file doesn't contain anything about how to run my `prefect-test.py` file. For this reason I created another deploy file in the root of my project with this:

```python
from prefect import flow
from prefect.docker import DockerImage

@flow(log_prints=True)
def buy():
    print("Selling securities")

if __name__ == "__main__":
    buy.deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
        image=DockerImage(
            name="docker.io/xxx/image",
            dockerfile="Dockerfile"
        ), 
        push=False
    )
```

I used dockerfile="Dockerfile" because I want Prefect to use my Dockerfile. If I remove that line and try to make a deploy, as I wrote above, Prefect tells me that I already have a Dockerfile in that folder. With that line, I get no error.
I don't want prefect create another build and push it, but I want it uses the image I already have on dockerhub. For this reason I put also:

`push=False`

If I deploy it and run the task, I get error:

```
timestamp,level,flow_run_id,task_run_id,message
	2024-12-01 15:06:40.326981+00:00,20,b89520d8-a1ca-4802-b2f0-c998e81c6029,,Worker 'DockerWorker e7ea2375-4f9b-4a61-a31e-a81e2b9c3ce0' submitting flow run 'b89520d8-a1ca-4802-b2f0-c998e81c6029'
	2024-12-01 15:06:41.402569+00:00,20,b89520d8-a1ca-4802-b2f0-c998e81c6029,,Completed submission of flow run 'b89520d8-a1ca-4802-b2f0-c998e81c6029'
	2024-12-01 15:06:44.210221+00:00,20,b89520d8-a1ca-4802-b2f0-c998e81c6029,,Opening process...
	2024-12-01 15:06:46.943062+00:00,20,b89520d8-a1ca-4802-b2f0-c998e81c6029,,Downloading flow code from storage at '.'
	2024-12-01 15:06:47.181580+00:00,40,b89520d8-a1ca-4802-b2f0-c998e81c6029,,"Flow could not be retrieved from deployment.
	Traceback (most recent call last):
	File ""<frozen importlib._bootstrap_external>"", line 850, in exec_module
	File ""<frozen importlib._bootstrap>"", line 228, in _call_with_frames_removed
	File ""deploy.py"", line 2, in <module>
	from prefect.docker import DockerImage
	ImportError: cannot import name 'DockerImage' from 'prefect.utilities.dockerutils' (/usr/local/lib/python3.9/site-packages/prefect/utilities/dockerutils.py)

	The above exception was the direct cause of the following exception:

	Traceback (most recent call last):
	File ""/usr/local/lib/python3.9/site-packages/prefect/engine.py"", line 414, in retrieve_flow_then_begin_flow_run
	flow = await load_flow_from_flow_run(flow_run, client=client)
	File ""/usr/local/lib/python3.9/site-packages/prefect/client/utilities.py"", line 51, in with_injected_client
	return await fn(*args, **kwargs)
	File ""/usr/local/lib/python3.9/site-packages/prefect/deployments/deployments.py"", line 264, in load_flow_from_flow_run
	flow = await run_sync_in_worker_thread(load_flow_from_entrypoint, str(import_path))
	File ""/usr/local/lib/python3.9/site-packages/prefect/utilities/asyncutils.py"", line 91, in run_sync_in_worker_thread
	return await anyio.to_thread.run_sync(
	File ""/usr/local/lib/python3.9/site-packages/anyio/to_thread.py"", line 33, in run_sync
	return await get_asynclib().run_sync_in_worker_thread(
	File ""/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py"", line 877, in run_sync_in_worker_thread
	return await future
	File ""/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py"", line 807, in run
	result = context.run(func, *args)
	File ""/usr/local/lib/python3.9/site-packages/prefect/flows.py"", line 1539, in load_flow_from_entrypoint
	flow = import_object(entrypoint)
	File ""/usr/local/lib/python3.9/site-packages/prefect/utilities/importtools.py"", line 201, in import_object
	module = load_script_as_module(script_path)
	File ""/usr/local/lib/python3.9/site-packages/prefect/utilities/importtools.py"", line 164, in load_script_as_module
	raise ScriptError(user_exc=exc, path=path) from exc
	prefect.exceptions.ScriptError: Script at 'deploy.py' encountered an exception: ImportError(""cannot import name 'DockerImage' from 'prefect.utilities.dockerutils' (/usr/local/lib/python3.9/site-packages/prefect/utilities/dockerutils.py)"")"
	2024-12-01 15:06:47.600011+00:00,20,b89520d8-a1ca-4802-b2f0-c998e81c6029,,Process for flow run 'nimble-wildcat' exited cleanly.
```
This is strange because that import is the same of the one in the example on Prefect v3 docs.

Any ideas?

Thanks

**Comment by zzstoatzz** [comment]
hi @Lukem76!

I notice you say
> I don't want prefect create another build and push it, but I want it uses the image I already have on dockerhub. For this reason I put also:

If this is the case, all you need to do is update your `.deploy()` call like this
```python
if __name__ == "__main__":
    buy.deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
        job_variables={"image": "your-registry/your-image:your-tag"},
        push=False
    )
```

to tell prefect use your `image` that you already have built on dockerhub, no `DockerImage` class required

### if you do want to build your own image, keep reading

i am not yet sure about the `ImportError` you are seeing, but i have produced a self-contained, working example that will hopefully help you [here](https://github.com/zzstoatzz/prefect-pack/blob/main/examples/run_flows_on_docker/buy.py), which uses [this `Dockerfile`](https://github.com/zzstoatzz/prefect-pack/blob/main/examples/run_flows_on_docker/Dockerfile).
```Dockerfile
FROM prefecthq/prefect:3-latest

WORKDIR /opt/prefect

COPY examples/run_flows_on_docker/buy.py examples/run_flows_on_docker/buy.py
```

There are numerous valid ways to write your dockerfile, but I did it like this because I was running my script from the root of the [repo](https://github.com/zzstoatzz/prefect-pack) and wanted to bake my flow into the image to reproduce your situation
```bash
» python examples/run_flows_on_docker/buy.py
```

**Reply by zzstoatzz** [reply]
let me know if this is helpful or if there are additional things I can help with here!

**Reply by Lukem76** [reply]
Hi zzstoatzz and thanks.

I was trying to use DockerImage import because I was running deploy from the root of my project where I already have  Dockerfile. Running it without using DockerImage where I can use dockerfile="Dockerfile", gives me the error "You already have a Dockerfile" or something similar. 
And if I try to use dockerfile="Dockerfile" without using DockerImage it gives me error because syntax is not correct.
Is it a correct approach to move the deploy file under another folder where the deploy can create its own temporary Dockerfile without the problem of finding an already existent one?
Thanks

**Reply by Lukem76** [reply]
I tried also to change deploy file like this:
```
`from prefect import flow
import subprocess

@flow
def buy():
    subprocess.run(["python", "prefect-test.py"])

if __name__ == "__main__":
    buy.deploy(
        name="yyy",
        work_pool_name="yyy",
        image="docker.io/xxx/image",
        push=False
    )`
```

It seems to work now.
but I had to add subprocess.run(["python", "prefect-test.py"]) in def buy.

Is this the correct approach?

**Reply by Lukem76** [reply]
Maybe I'm missing something because if I delete my project Dockerfile and put this deploy file:

```
from prefect import flow
import subprocess

@flow
def buy():
    subprocess.run(["python", "prefect-test.py"])

if __name__ == "__main__":
    buy.deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
        image="docker.io/user/repo:image",
        push=False
    )
```

in the root of my project and run python deploy and finally I run the task, it works

But if I leave my Dockerfile in the root and move the deploy file in a subfolder and run python deploy from there, it takes only few seconds to finish (while before it took longer), and when I run the task, it does nothing. It seems that the final image is empty this way

I don't understand which is the right way to do this. 
I think also doc is a little bit confusing about this.

I don't want to delete my dockerfile because it's a separate project that could be run on another environment. 
What am I missing?

P.S.: using job_variables={"image": "your-registry/your-image:your-tag"} it gives me error of "image" directive missing

Thanks

**Reply by zzstoatzz** [reply]
hi @Lukem76 - sorry if I confused with my example, but you generally shouldn't have to use `subprocess.run` to invoke prefect flows - I was just using it to create a self-contained MRE of your situation. Can you explain why you are using `subprocess.run` above?

I would generally recommend writing your own `Dockerfile`, you can use the same one for many deployments if you want, or define one per project (i recommend the latter)

in addition to the [docs](https://docs.prefect.io/v3/deploy/infrastructure-examples/docker), there is an example [here](https://github.com/zzstoatzz/prefect-pack/blob/main/flows/with_special_deps/deploy.py) which should be roughly what I'd recommend.

> P.S.: using job_variables={"image": "your-registry/your-image:your-tag"} it gives me error of "image" directive missing

It's easiest to help with actual stack traces and the code you used to produce that stack trace

**Reply by Lukem76** [reply]
Hi @zzstoatzz 

Sorry if I was not clear in my explanation but I'm new to prefect and I'm really interested in using it.

This is my situation:
I have an already containerizer project that is a simple python file "prefect-test.py" that print a simple message (this is only an example, I could have different projects like this that do different things).
Its Dockerfile is:
```
FROM python:3.9-slim

WORKDIR /app

COPY . .

RUN pip install --no-cache-dir -r requirements.txt

CMD ["python", "prefect-test.py"]
```

I need to schedule a job in Prefect v3 in order to run it every hour. I want Prefect to run my container, I don't want to change my python file to adapt it to Prefect (adding decorators and so on...). My original project can't be changed.

What I can do is to create a deploy file with something that can execute my container inside Prefect.

How do I have to write my deploy file exactly?
Using a deploy file like this
```
from prefect import flow
import subprocess

@flow
def buy():
    WHAT DO I HAVE TO PLACE HERE???

if __name__ == "__main__":
    buy.deploy(
        name="my-deployment",
        work_pool_name="my-pool",
        image="xxx/yyy",
        push=False
    )
```
It's not clear for me what to insert inside "def buy():"
I tried with subprocess.run passing "prefect-test.py" because chatGPT suggested me :).
Do I have to insert the name of a function inside prefect-test.py file?

Using subprocess.run(["python", "prefect-test.py"]), deleting my project Dockerfile and running python deploy.py placed in the root, the deployment seems to work because I can run it and see the log printed by my prefect-test.py.

As I wrote before, If I create a deploy.py file inside the root of my project without deleting my project Dockerfile and run "python deploy.py", it tells me that I have already a Dockerfile (the file of my original project)

Using in the deploy.py file something like this:
```
image=DockerImage(
            name="my_image",
            tag="deploy-guide",
            dockerfile="Dockerfile"
)
```
forcing this way to use my original project Dockerfile, then it gives me an error:
prefect.exceptions.ScriptError: Script at 'deploy.py' encountered an exception: ImportError("cannot import name 'DockerImage' from 'prefect.utilities.dockerutils' (/usr/local/lib/python3.9/site-packages/prefect/utilities/dockerutils.py)").

I also tried to update python version in my image, using 3.11 but the error is the same.

In the github ecample you lined me there are too many examplese an I don't know whato to see.

I wrote you all the code I currently have except prefect-test.py because it only print a simple message. I don't have anything else to show you.

Thanks
